pandas/tests/base/test_ops.py

```diff
@@ -137,227 +137,238 @@ def setup_method(self, method):
-                assert result[0] == orig[0]
-                for r in result:
-                    assert isinstance(r, Timestamp)
-
-                tm.assert_numpy_array_equal(
-                    result.astype(object), orig._values.astype(object)
-                )
-            else:
-                tm.assert_numpy_array_equal(result, orig.values)
-                assert result.dtype == orig.dtype
+        result = obj.value_counts()
+        tm.assert_series_equal(result, expected_s)
+        assert result.index.name is None
+
+        result = obj.unique()
+        if isinstance(obj, Index):
+            assert isinstance(result, type(obj))
+            tm.assert_index_equal(result, orig)
+            assert result.dtype == orig.dtype
+        elif is_datetime64tz_dtype(obj):
+            # datetimetz Series returns array of Timestamp
+            assert result[0] == orig[0]
+            for r in result:
+                assert isinstance(r, Timestamp)
+
+            tm.assert_numpy_array_equal(
+                result.astype(object), orig._values.astype(object)
+            )
+        else:
+            tm.assert_numpy_array_equal(result, orig.values)
+            assert result.dtype == orig.dtype
 
-            assert o.nunique() == len(np.unique(o.values))
+        # dropna=True would break for MultiIndex
+        assert obj.nunique(dropna=False) == len(np.unique(obj.values))
 
     @pytest.mark.parametrize("null_obj", [np.nan, None])
-    def test_value_counts_unique_nunique_null(self, null_obj):
-
-        for orig in self.objs:
-            o = orig.copy()
-            klass = type(o)
-            values = o._ndarray_values
-
-            if not allow_na_ops(o):
-                continue
-
-            # special assign to the numpy array
-            if is_datetime64tz_dtype(o):
-                if isinstance(o, DatetimeIndex):
-                    v = o.asi8
-                    v[0:2] = iNaT
-                    values = o._shallow_copy(v)
-                else:
-                    o = o.copy()
-                    o[0:2] = pd.NaT
-                    values = o._values
-
-            elif needs_i8_conversion(o):
-                values[0:2] = iNaT
-                values = o._shallow_copy(values)
+    def test_value_counts_unique_nunique_null(self, null_obj, index_or_series_obj):
+        orig = index_or_series_obj
+        obj = orig.copy()
+        klass = type(obj)
+        values = obj._ndarray_values
+        num_values = len(orig)
+
+        if not allow_na_ops(obj):
+            pytest.skip("type doesn't allow for NA operations")
+        elif isinstance(orig, (pd.CategoricalIndex, pd.IntervalIndex)):
+            pytest.skip(f"values of {klass} cannot be changed")
+        elif isinstance(orig, pd.MultiIndex):
+            pytest.skip("MultiIndex doesn't support isna")
+
+        # special assign to the numpy array
+        if is_datetime64tz_dtype(obj):
+            if isinstance(obj, DatetimeIndex):
```

Comment:
this can be simplified; the else path *should* work for all

Reply:
xref: #32205